File: ValidateBST1.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Mon Dec 16 12:12:31 2019

@author: tanvirkaur
"""
# leetcode Acceptance = Yes
# Time complexity = O(n)
# Space complexity = O(1) if we donot consider the recursion stack
class Solution(object):
    def isValidBST(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        return self.minmax(root, float('-inf'), float('inf'))
    
    def minmax(self, root, minn, maxx):
        # base case
        if not root:
            return True
        if minn!= None and root.val <= minn:
            return False
        if maxx!= None and root.val >= maxx:
            return False
        # logic
        left= self.minmax(root.left, minn, root.val) 
        right = self.minmax(root.right, root.val, maxx)
        return left and right

File: ValidateBST1.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
/*
TC: O(n)
SC: O(h)
*/
class Solution {
    public boolean isValidBST(TreeNode root) {
        if (root == null) {
            return true;
        }

        return isValidBST(root, null, null);
    }

    private boolean isValidBST(TreeNode node, Integer min, Integer max) {
        if (node == null) return true;

        if ((min != null && node.val <= min) || (max!= null && node.val >= max)) {
            return false;
        }

        return isValidBST(node.left, min, node.val) && isValidBST(node.right, node.val, max);
    }
}